proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,717 @@
+  hlsl::RootSignature::RootParameterType ParameterType;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+  std::variant<RootSignature::ContainerRootConstants,
+                   RootSignature::ContainerRootDescriptor,
+                   RootSignature::ContainerRootDescriptorTable> Extra;
+};
+
+struct StaticSamplerDesc {
+  hlsl::RootSignature::Filter Filter =
+      hlsl::RootSignature::Filter::ANISOTROPIC;
+  hlsl::RootSignature::TextureAddressMode AddressU =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  hlsl::RootSignature::TextureAddressMode AddressV =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  hlsl::RootSignature::TextureAddressMode AddressW =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  float MipLODBias = 0;
+  uint32_t MaxAnisotropy = 16;
+  hlsl::RootSignature::ComparisonFunc ComparisonFunc =
+      hlsl::RootSignature::ComparisonFunc::LessEqual;
+  hlsl::RootSignature::StaticBorderColor BorderColor =
+      hlsl::RootSignature::StaticBorderColor::OpaqueWhite;
+  float MinLOD = 0;
+  float MaxLOD = hlsl::RootSignature::Float32Max;
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+};
+
+struct ParsedRootSignature {
+  RootSignature::RootSignatureVersion Version;
+  uint32_t RootFlags;
+  llvm::SmallVector<RootSignature::ContainerRootParameter, 8>
+      RSParameters;
+  llvm::SmallVector<RootSignature::StaticSamplerDesc, 8> StaticSamplers;
+};
+
+}
+
+```
+
+This will require more work for print and serialization the AST node.
+
+The AST node serialization code might be looked like this:
+```
+void ASTDeclWriter::VisitHLSLRootSignatureDecl(HLSLRootSignatureDecl *D) {
+  VisitNamedDecl(D);
+  Record.AddSourceLocation(D->getKeywordLoc());
+  hlsl::ParsedRootSignature &RS = D->getRootSignature();
+  Record.push_back(RS.RSDesc.Version);
+  Record.push_back(RS.RSDesc.Flags);
+  for (const auto &P : ParsedRS.RSParameters) {
+    // push_back all the fields of root parameters.
+  }
+  for (const auto &SS : ParsedRS.StaticSamplers) {
+    // push_back all the fields of static samplers.
+  }
+  for (const auto &Range : ParsedRS.DescriptorRanges) {
+    // push_back all the descriptor ranges.
+  }
+  
+  Code = serialization::DECL_HLSL_ROOT_SIGNATURE;
+}
+
+```
+
+
+A HLSLRootSignatureAttr will be created when meet RootSignature attribute in
+HLSL.
+
+```
+
+[RootSignature(MyRS1)]
+float4 main(float4 coord : COORD) : SV_Target
+{
+…
+}
+
+```
+
+Because the RootSignature attribute in hlsl only have the string, it is easier
+to add a StringArgument to save the string first.
+A HLSLRootSignatureDecl will be created and added to the HLSLRootSignatureAttr 
+for diagnostic and saved to the HLSLEntryRootSignatureAttr for clang 
+code generation.
+The HLSLRootSignatureDecl will save StringLiteral instead StringRef for 
+diagnostic.
```

Comment:
I don't understand the significance of using StringLiteral instead of StringRef here.  Is there something about StringLiteral that makes diagnostics easier?  What's the reason for the decision to use one over the other?

Reply:
StringLiteral has source location which will help to emit diagnostic.